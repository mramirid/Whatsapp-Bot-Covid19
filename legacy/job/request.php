<?php

require_once 'functions/covid_id.php';
require_once 'functions/covid_prov.php';

date_default_timezone_set('Asia/Jakarta');


/* ---------------------- Data Nasional ---------------------- */

// Ambil data terakhir di tabel nasional
$dataDBNasional = CovidID\getTodayData();

// Ambil statistik kasus di Indonesia
$dataApiNasional = CovidID\fetchUpdateStatistik();

if (!(array) $dataDBNasional) :
    // Jika data kosong, lakukan insert
    CovidID\insertNewRowToday($dataApiNasional);
elseif (CovidID\isDBExpired($dataDBNasional, $dataApiNasional)) :
    // Jika data hari ini sudah usang, update
    CovidID\updateTodayData($dataDBNasional->id, $dataApiNasional);
endif;

/* ---------------------- End of Data Nasional ---------------------- */


/* ---------------------- Data Provinsi ---------------------- */

// Ambil data terakhir di tabel pengambilan_provinsi & detail_penambilang_provinsi
$dataDBProvinsi = CovidProv\getTodayData();

// Ambil statistik kasus di tiap provinsi
$dataApiProvinsi = CovidProv\fetchUpdateStatistik();

// Ambil data terakhir di tabel pengambilan_provinsi
$dataDBProvinsi_Peng = CovidProv\getTodayData_Peng();

if ($dataDBProvinsi->num_rows == 0) :
    CovidProv\insertNewDataToday($dataApiProvinsi);
elseif (\CovidProv\isDBExpired($dataDBProvinsi, $dataApiProvinsi, $dataDBProvinsi_Peng)) :
    CovidProv\updateTodayData($dataApiProvinsi);
endif;

/* ---------------------- End of Data Provinsi ---------------------- */
