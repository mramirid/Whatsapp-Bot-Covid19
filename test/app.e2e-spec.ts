import { CACHE_MANAGER } from '@nestjs/cache-manager';
import type { INestApplication } from '@nestjs/common';
import { Test } from '@nestjs/testing';
import request from 'supertest';

import AppController from '../src/app.controller';
import AppService from '../src/app.service';
import dummyCountryStats from '../src/country/country-stats.dummy';
import CountryService from '../src/country/country.service';
import UpstreamAPI from '../src/upstream-api/upstream-api.abstract';

describe('AppController (e2e)', () => {
  let app: INestApplication;

  beforeEach(async () => {
    const cacheManagerMock = {
      get: async () => dummyCountryStats,
    };

    const moduleFixture = await Test.createTestingModule({
      controllers: [AppController],
      providers: [
        AppService,
        CountryService,
        { provide: CACHE_MANAGER, useValue: cacheManagerMock },
        { provide: UpstreamAPI, useValue: {} },
      ],
    }).compile();

    app = moduleFixture.createNestApplication();
    await app.init();
  });

  it('should reply unknown command with the correct WhatsApp message', () => {
    return request(app.getHttpServer())
      .post('/')
      .send({
        ProfileName: 'Amir',
        Body: 'Test',
      })
      .expect(200)
      .expect((res) => {
        expect(res.text).toMatchSnapshot();
      });
  });

  it('should reply /halo command with the correct WhatsApp message', () => {
    return request(app.getHttpServer())
      .post('/')
      .send({
        ProfileName: 'Amir',
        Body: '/halo',
      })
      .expect(200)
      .expect((res) => {
        expect(res.text).toMatchSnapshot();
      });
  });

  it('should reply /help command with the correct WhatsApp message', () => {
    return request(app.getHttpServer())
      .post('/')
      .send({
        ProfileName: 'Amir',
        Body: '/help',
      })
      .expect(200)
      .expect((res) => {
        expect(res.text).toMatchSnapshot();
      });
  });

  it('should reply /nasional command with the correct WhatsApp message', () => {
    return request(app.getHttpServer())
      .post('/')
      .send({
        ProfileName: 'Amir',
        Body: '/nasional',
      })
      .expect(200)
      .expect((res) => {
        expect(res.text).toMatchSnapshot();
      });
  });
});
