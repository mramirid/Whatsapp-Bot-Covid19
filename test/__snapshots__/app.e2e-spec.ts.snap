// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppController (e2e) should reply /halo command with the correct WhatsApp message 1`] = `
"<?xml version="1.0" encoding="UTF-8"?><Response><Message>Halo Amir,

Saya adalah bot yang didesain untuk memberikan info kasus COVID-19 di Indonesia.
Sumber data didapatkan dari API yang disediakan oleh Open Disease Data API (https://corona.lmao.ninja/).
Ketikkan \`\`\`/help\`\`\` untuk mendapatkan informasi mengenai saya dan daftar perintah yang tersedia.</Message></Response>"
`;

exports[`AppController (e2e) should reply /help command with the correct WhatsApp message 1`] = `
"<?xml version="1.0" encoding="UTF-8"?><Response><Message>Daftar perintah:
1. \`\`\`/halo\`\`\` -&gt; Perkenalan bot
2. \`\`\`/nasional\`\`\` -&gt; Statistik COVID-19 di Indonesia
3. \`\`\`/help\`\`\` -&gt; Daftar perintah

_Versi 1.0 - 17/05/2020 20.00: Bot dibuat._
_Versi 2.0 - 13/05/2023 10.06: Penggantian API &amp; penghapusan fitur statistik per-provinsi._</Message></Response>"
`;

exports[`AppController (e2e) should reply /nasional command with the correct WhatsApp message 1`] = `
"<?xml version="1.0" encoding="UTF-8"?><Response><Message>*Statistik COVID-19 di Indonesia*

Total: 131.524.885 (+45.442.552)
Sembuh: 90.332 (+7.223)
Meninggal: 3.352.999 (+823.993)
Dirawat: 583
Kritis: 2

Tetap jaga kesehatan dan patuhi protokol kesehatan.

_Pembaharuan terakhir pada 14 Mei 2023, 08.24.30 WIB._</Message></Response>"
`;

exports[`AppController (e2e) should reply unknown command with the correct WhatsApp message 1`] = `"<?xml version="1.0" encoding="UTF-8"?><Response><Message>Maaf perintah tidak dikenali, kirimkan saya \`\`\`/help\`\`\` untuk melihat daftar perintah.</Message></Response>"`;
